File: Rust/74. Search a 2D Matrix.rs
use std::cmp::Ordering;

impl Solution {
    pub fn search_matrix(matrix: Vec<Vec<i32>>, target: i32) -> bool {
        let m = matrix.len();
        let n = matrix[0].len();

        let mut row = 0;
        while row < m {
            if matrix[row][n - 1] >= target {
                break;
            }
            row += 1;
        }

        if row < m {
            let mut low = 0;
            let mut high = n as isize - 1;

            while low <= high {
                let mid = low + (high - low) / 2;
                match matrix[row][mid as usize].cmp(&target) {
                    Ordering::Equal => return true,
                    Ordering::Less => low = mid + 1,
                    Ordering::Greater => high = mid - 1,
                }
            }
        }

        false
    }
}

File: Java/74. Search a 2D Matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int m = matrix.length, n = matrix[0].length;
        int row = 0;

        while (row < m) {
            if (matrix[row][n - 1] >= target)
                break;
            row++;
        }

        if (row < m) {
            int left = 0, right = n - 1;

            while (left <= right) {
                int mid = left + (right - left) / 2;
                if (matrix[row][mid] == target)
                    return true;

                if (matrix[row][mid] < target)
                    left = mid + 1;
                else
                    right = mid - 1;
            }
        }

        return false;
    }
}